src/librustc/ty/walk.rs

```diff
@@ -92,11 +92,19 @@ fn push_subtypes<'tcx>(stack: &mut TypeWalkerStack<'tcx>, parent_ty: Ty<'tcx>) {
         ty::TyProjection(ref data) => {
             stack.extend(data.trait_ref.substs.types().rev());
         }
-        ty::TyTrait(ref obj) => {
-            stack.extend(obj.principal.input_types().rev());
-            stack.extend(obj.projection_bounds.iter().map(|pred| {
-                pred.0.ty
-            }).rev());
+        ty::TyDynamic(ref obj, ..) => {
+            stack.extend(obj.iter().rev().flat_map(|predicate| {
+                let types = match *predicate.skip_binder() {
+                    ty::ExistentialPredicate::Trait(tr) => tr.substs.types().rev().chain(None),
+                    ty::ExistentialPredicate::Projection(p) =>
+                        p.trait_ref.substs.types().rev().chain(Some(p.ty)),
+                    ty::ExistentialPredicate::AutoTrait(_) =>
+                        // Empty iterator
+                        ty::Substs::empty().types().rev().chain(None),
+                };
+
+                types.rev()
```

Comment:
This is reversing the already reversed types, it should be just the chained iterator, no extra `.rev()`.